tq/meter.go

```diff
@@ -86,44 +80,108 @@ func WithOSEnv(os env) meterOption {
 	return WithLogFile(name)
 }
 
-// NewMeter creates a new ProgressMeter.
-func NewMeter(options ...meterOption) *ProgressMeter {
-	m := &ProgressMeter{
-		logger:         &progressLogger{},
+type MeterOption struct {
+	// DryRun is an option that determines whether updates should be sent to
+	// stdout.
+	DryRun bool
+	// LogFile is an option that sends updates to a text file.
+	LogFile string
+	// OS is an option that sends updates to the text file path specified in
+	// the OS Env.
+	OS interface {
+		Get(key string) (val string, ok bool)
+	}
+}
+
+func (o *MeterOption) configure(m *Meter) {
+	m.dryRun = o.DryRun
+
+	if f, ok := o.logFile(); ok {
+		m.logToFile = 1
+		m.logger = tools.NewSyncWriter(f)
+	}
+}
+
+func (o *MeterOption) logFile() (*os.File, bool) {
+	var name string = o.LogFile
+	if len(name) == 0 {
+		name, _ = o.OS.Get("GIT_LFS_PROGRESS")
+		if len(name) == 0 {
+			return nil, false
+		}
+	}
+
+	printErr := func(err string) {
+		fmt.Fprintf(os.Stderr, "Error creating progress logger: %s\n", err)
+	}
+
+	if !filepath.IsAbs(name) {
+		printErr("GIT_LFS_PROGRESS must be an absolute path")
+		return nil, false
+	}
+
+	if err := os.MkdirAll(filepath.Dir(name), 0755); err != nil {
+		printErr(err.Error())
+		return nil, false
+	}
+
+	file, err := os.OpenFile(name, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0666)
+	if err != nil {
+		printErr(err.Error())
+		return nil, false
+	}
+
+	return file, true
+}
+
+// NewMeter creates a new Meter.
+func NewMeter(opt *MeterOption) *Meter {
```

Comment:
Do we really need the `MeterOption` type? The only two things to change are whether dry run is enabled, and what logger is used.

```go
m := NewMeter()
m.DryRun = true
m.Logger = tq.LoggerToFile(filename)
m.Logger = tq.LoggerFromEnv(osEnv)
m.Start()
```

My hunch is that we didn't use fields like this to prevent race conditions. Should we continue trying to protect the caller, or can we simplify things and just document that those fields should be set before `Start()` is called? FWIW I do prefer the `MeterOptions` type to the option func args :+1: